CLI tools for installing/testing Docker and Singularity containers

simonbray opened the pull request:
Previously in a PR to the galaxy-lib repo: https://github.com/galaxyproject/galaxy-lib/pull/87

simonbray committed:
add work from https://github.com/galaxyproject/galaxy-lib/pull/87

simonbray committed:
Update .travis.yml

simonbray committed:
Update .travis.yml

mvdbeek commented on .travis.yml:
```diff
@@ -1,11 +1,30 @@
 
+before_install:
+  # Install Singularity
+  - sudo apt install -y build-essential libssl-dev uuid-dev libgpgme11-dev squashfs-tools wget git
```
We are not using travis for these tests anymore. The docker testing image comes pre-installed with singularity. For a simple test I'd suggest you create something unit-test like in `test/integration`, which will be picked up automatically.

mvdbeek commented on lib/galaxy/tools/deps/mulled/get_tests.py:
```diff
@@ -0,0 +1,304 @@
+    return package_tests
+
+
+doctest.testmod()
```
I'd suggest avoiding doctests. We're currently using pytest, but you can also use unittest tests if you like.

mvdbeek commented on lib/galaxy/tools/deps/mulled/mulled_list.py:
```diff
@@ -0,0 +1,168 @@
+    # main()
+
+    import doctest
+    doctest.testmod()
```
Again, those doctests should probably live as unit tests in test/integration, or if you want, test/mulled

mvdbeek commented:
Let me know if you'd like me to re-create some of the doctests as a unit test

simonbray commented:
Thanks for the offer; I'll let you know if I need a hand.

simonbray commented:
@mvdbeek would you be able to convert the doctests? Or perhaps at least a few examples so I have an idea of how to do the others.

mvdbeek commented:
I'll try to get something started this week

simonbray commented:
Thanks! Much appreciated.

mvdbeek committed:
Convert 2 mulled-cli doctests to unit-test style test

We may or may not want add these as integration or unit tests.
It probably makes a lot of sense to keep these as unit tests
and skip when singularity is not available.

mvdbeek commented:
@simonbray I converted 2 tests to unittest style tests here: https://github.com/simonbray/galaxy/pull/1

simonbray committed:
add unit-tests for mulled tools

simonbray committed:
revert changes to .travis.yml

simonbray committed:
move scripts to new location

simonbray committed:
add jinja2 to reqs; fix python3 error downloading tests

simonbray commented:
Hi @mvdbeek, I've got most of the tests passing now, with one exception. This PR needs the jinja2 library and therefore I added it to `pinned-requirements.txt` in the most recent commit. However, the test_galaxy_packages test still fails with `ImportError: No module named 'jinja2'`. Do you have an idea why?

mvdbeek commented:
You'd have to add jinja2 to `packages/tool_util/requirements.txt`

simonbray committed:
add jinja2 to packages/tool_util/requirements.txt

jmchilton committed:
Improvements to mulled functionality testing from #7936.

- No fixed paths.
- Annotate unit tests as slow and requiring external network access so default suite stays fast.

jmchilton committed:
Sync dev and production jinja requirement.

jmchilton committed:
Lint fixes (would rebase if it were my branch)

simonbray commented:
@jmchilton, I'm happy to finish this, but would really appreciate some guidance about how much of this PR is useful or in scope. For example, we already used some of the code here: https://github.com/BioContainers/singularity-build-bot; I don't know if you think it's worth duplicating it in the Galaxy codebase?

Thanks!

jmchilton committed:
Fix up galaxy-tool-util package structure for #7936.

mvdbeek commented:
@galaxybot test this

mvdbeek commented:
Awesome, thanks a lot @simonbray!

simonbray commented:
Oh, I missed this. Thanks for the merge @mvdbeek!

Should I now make a PR to the galaxy-lib repo (which was the original goal)? Or will this be done automatically?

mvdbeek commented:
galaxy-lib will be phased out, the mulled utils live in galaxy-tool-utils. You should be able to install galaxy-tool-util (`pip install galaxy-tool-util`) and that should place the mulled scripts on your PATH.
